craftdb.lua

```diff
@@ -214,28 +226,125 @@ function CraftDB:get_all_recipes(item_list)
 end
 
 
--- TODO: Add filtering so we can exclude nodes made via tablesaw.
-function CraftDB:find_all_matching_items(name_pattern, offset, max_count)
+function CraftDB:search_items(name_pattern, options)
   local matching_names = {}  -- list of names
-  local image_names = {}  -- name -> image_filename
 
-  if type(name_pattern) ~= 'string' then name_pattern = '.' end
+  if type(name_pattern) ~= 'string' then return {} end
+
+  if type(options) ~= 'table' then options = {} end
+
+  -- 'offset' must be an integer and >= 1.
+  -- If not, set `offset` = 1.
+  local offset = (type(options['offset']) == 'number') and
+                 math.max(1, math.floor(options['offset'])) or 1
+
+  -- 'max_count' must be an integer, and 1 <= max_count <= MAX_MATCHES.
+  -- If not, set `max_count` = MAX_MATCHES.
+  local max_count = (type(options['max_count']) == 'number') and
+              math.min(math.max(1, math.floor(options['max_count'])),
+                       MAX_MATCHES) or
+              MAX_MATCHES
+
+  -- 'group_filter' might be nil, but if present, should always be a table.
+  local group_filter = options['group_filter']
+  if (type(group_filter) ~= 'table') and (type(group_filter) ~= nil) then
+    group_filter = {}
+  end
+
+  -- Generate exclusion lookup tables from the input lists.
+  local exclude_mods = _make_lut_from_list(options['exclude_mods'])
+
+  -- Are we looking up a group or item?  Groups start with 'group:'
+  local group = nil
+  if name_pattern:sub(1, 6) == 'group:' then
+    group = name_pattern:sub(7)
+    name_pattern = nil
+  end
+
+  -- Internal helper method, uses lambda capture of 'name_pattern'.
+  -- NOTE: `string.find()` treats '.' as a meta-character that will match
+  -- anything.
+  -- Should return a bool.
+  local function is_name_match(name)
+    if options['substring_match'] then
+      return nil ~= string.find(name, name_pattern)
```

Comment:
Switching to find alone wont work but find can disable pattern matching, basically this should be `string.find(name, name_pattern, nil, true)` (`nil` is starting point, can be `1` as well for first character)
Only if fourth argument is `true` it will treat  second argument as plain text instead of lua pattern.

(Also after setting fourth to true you can remove this note: <code>-- NOTE: \`string.find()\` treats '.' as a meta-character that will match</code>)

This is same problem that `string.match`, pattern matching must be disabled.
For very likely and simple example: with current code search for `"("` and your game will crash.

Reply:
Ok.  4th param is now true, '.' matches nothing (instead of everything), and verified that '' and ':' will match everything.  I also added unit tests so ensure that "(" and ")" won't cause mod to crash.